pkg/devfile/adapters/docker/component/adapter.go

```diff
@@ -0,0 +1,219 @@
+	componentName := a.ComponentName
+
+	// Get the project source volume
+	volumeLabels := map[string]string{
+		"component": componentName,
+		"type":      "projects",
+	}
+	vols, err := a.Client.GetVolumesByLabel(volumeLabels)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to retrieve source volume for component "+componentName)
+	}
+	if len(vols) == 0 {
+		return fmt.Errorf("Unable to find source volume for component %s", componentName)
+	}
+	projectVolumeName := vols[0].Name
+
+	supportedComponents := adaptersCommon.GetSupportedComponents(a.Devfile.Data)
+	if len(supportedComponents) == 0 {
+		return fmt.Errorf("No valid components found in the devfile")
+	}
+
+	for _, comp := range supportedComponents {
+		// Check to see if this component is already running and if so, update it
+		// If component isn't running, re-create it, as it either may be new, or crashed.
+		containers, err := a.Client.GetContainersByComponentAndAlias(componentName, *comp.Alias)
+		if err != nil {
+			return errors.Wrapf(err, "unable to list containers for component %s", componentName)
+		}
+		if len(containers) == 0 {
+			// Container doesn't exist, so need to pull its image (to be safe) and start a new container
+			s := log.Spinner("Pulling image " + *comp.Image)
+			defer s.End(false)
+
+			err = a.Client.PullImage(*comp.Image)
+			if err != nil {
+				return errors.Wrapf(err, "Unable to pull %s image", *comp.Image)
+			}
+			s.End(true)
+
+			// Start the container
+			err = a.startContainer(componentName, projectVolumeName, comp)
+			if err != nil {
+				return errors.Wrapf(err, "Unable to start container for devfile component %s", *comp.Alias)
+			}
+			glog.V(3).Infof("Successfully created container %s for component %s", *comp.Image, componentName)
+		} else {
+			// Container already exists
+			containerID := containers[0].ID
+
+			// Get the associated container config from the container ID
+			containerConfig, err := a.Client.GetContainerConfig(containerID)
+			if err != nil {
+				return errors.Wrapf(err, "unable to get the container config for component %s", componentName)
+			}
+
+			// See if the container needs to be updated
+			if utils.DoesContainerNeedUpdating(comp, containerConfig) {
+				glog.V(3).Info("Updating the component")
+				// Remove the container
+				err := a.Client.RemoveContainer(containerID)
+				if err != nil {
+					return errors.Wrapf(err, "Unable to remove container %s for component %s", containerID, *comp.Alias)
+				}
+
+				// Start the container
+				err = a.startContainer(componentName, projectVolumeName, comp)
+				if err != nil {
+					return errors.Wrapf(err, "Unable to start container for devfile component %s", *comp.Alias)
+				}
+				glog.V(3).Infof("Successfully created container %s for component %s", *comp.Image, componentName)
+			}
+		}
+	}
+	return nil
+}
+
+func (a Adapter) startContainer(componentName string, projectVolumeName string, comp versionsCommon.DevfileComponent) error {
```

Comment:
Can this be moved to docker/utils/utils.go to be consistent with kubernetes package?

Reply:
Makes sense to me 👍

Reply:
Actually, this should stay in adapter.go since it relies on the Adapter struct defined in this file (and it's equivalent to `createOrUpdateComponent` in `kubernetes/component/adapter.go`)